src/cassiopeia/utils.py

```diff
@@ -180,3 +181,54 @@ def _combine_edge_data(parent_edge: dict[str, Any], child_edge: dict[str, Any])
         child_length = child_edge.get("length", 0)
         new_edge["length"] = parent_length + child_length
     return new_edge
+
+
+def _get_character_matrix(tree, characters_key: str = "characters") -> np.ndarray:
+    """Get character matrix from a tree object."""
+    if isinstance(tree, CassiopeiaTree):
+        if characters_key == "characters":
+            character_matrix = tree.character_matrix
+        else:
+            character_matrix = tree.layers[characters_key]
+    elif isinstance(tree, TreeData):
+        character_matrix = tree.obsm[characters_key]
+    return np.asarray(character_matrix)
+
+
+def _get_missing_state_indicator(tree: TreeLike) -> int:
+    """Get the missing state indicator from a tree object."""
+    if isinstance(tree, CassiopeiaTree):
+        return tree.missing_state_indicator
+    elif isinstance(tree, TreeData):
+        return tree.uns.get("missing_state_indicator", -1)
+
+
+def _get_tree_parameter(tree, param_name: str, default=None):
+    """Get a parameter from CassiopeiaTree or TreeData."""
+    if isinstance(tree, CassiopeiaTree):
+        return tree.parameters.get(param_name, default)
+    elif isinstance(tree, TreeData):
+        return tree.uns.get("cassiopeia_parameters", {}).get(param_name, default)
+    return default
+
+
+def get_mean_depth(tree: TreeLike, depth_key, tree_key: str | None = None) -> float:
```

Comment:
This function should probably generate a useful error message if depth_key is not present. Maybe add something similar to the _check_tree_has_key function in pycea? 

```
def _check_tree_has_key(tree: nx.DiGraph, key: str):
    """Checks that tree nodes have a given key."""
    # sample 10 nodes to check if the key is present
    sampled_nodes = random.sample(list(tree.nodes), min(10, len(tree.nodes)))
    for node in sampled_nodes:
        if key not in tree.nodes[node]:
            message = f"One or more nodes do not have {key} attribute."
            if key == "depth":
                message += " You can run `pycea.pp.add_depth` to add depth attribute."
            raise ValueError(message)
```